ext/standard/array.c

```diff
@@ -349,6 +350,65 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
 }
 /* }}} */
 
+/* return int to be compatible with compare_func_t */
+static int hash_zval_strict_function(zval *z1, zval *z2) /* {{{ */
+{
+	// The most important thing about this comparison mode is that the result
+	// is 0 when zend_is_identical, and non-zero otherwise.
+	if (zend_is_identical(z1, z2)) {
+		return 0;
+	}
+	// Below this point, the return value should always be non-zero.
+	// First try to use a "normal" comparison.
+	int c = zend_compare(z1, z2);
+	if (c != 0) {
+		return c;
+	}
+	// If the types are different, compare based on type.
+	int t1 = Z_TYPE_P(z1);
+	int t2 = Z_TYPE_P(z2);
+	if ( t1 != t2 ) {
+		return (t1 < t2 ) ? -1 : 1;
+	}
+	switch (t1) {
+	case IS_ARRAY:
+		// Do a recursive comparison
+		return zend_hash_compare(
+			Z_ARRVAL_P(z1), Z_ARRVAL_P(z2),
+			(compare_func_t) hash_zval_strict_function, 1 /* ordered */
+		);
+
+	case IS_OBJECT:
+	{
+		// First compare on class name, as that's consistent across runs
+		zend_string *name1 = Z_OBJCE_P(z1)->name;
+		zend_string *name2 = Z_OBJCE_P(z2)->name;
+		c = zend_binary_strcmp(name1->val, name1->len, name2->val, name2->len);
+		if (c != 0) {
+			return (c < 0) ? -1 : 1;
+		}
+		// Fall back on spl_object_id() value, which will probably vary
+		// non-deterministically between runs (alas).
+		zend_object *obj1 = Z_OBJ_P(z1);
+		zend_object *obj2 = Z_OBJ_P(z2);
+		ZEND_ASSERT(obj1->handle != obj2->handle);
+		return (obj1->handle < obj2->handle) ? -1 : 1;
+	}
+
+	default:
+		ZEND_ASSERT(0 && "Value types should have consistent == and ===");
+		return 0;
```

Comment:
This still doesn't cover everything, for example consider `"0"` and `"0.0"`. I would recommend to explicitly write out the comparison function and not rely on zend_compare at all.

Reply:
Ok, done.  Good advice.